### DISCUSSION TITLE
Creating favourite songs playlists for archiving

### DISCUSSION THREAD

**Discussion by ultrlight** [Chimildic/goofy#150] [Помощь с алгоритмом] [answered]
Hello, firstly I would like to thank you for this program! I am attempting to collect an archive of my Spotify favourites by making the following playlists :

1. A playlist that logs every song I have ever added to my favourite songs. 
2. A playlist that logs all the songs that I remove from my favourite songs.
3. A playlist that shows my favourite songs at present. 

I have managed to make Playlist 3 successfully. I would like advice on how to write the first two playlists in a simple way. Apologies if anything doesn't make sense, please let me know. :)

**Comment by Chimildic** [comment, marked as answer]
Set trigger to function `logFavoriteSongs` (is saving snapshot of present favorite tracks and append to last with dedup).
 In dependence from your activity choose period. If _every day_, tracks that liked and removed in one day dont push to cache. You can set an hour. But such fleeting tracks dont make sense. In case _day of week_ can lost many tracks that liked and removed before to create new snapshot.

Constant `COUNT_TRACKS` is how many last favorite tracks need to check. If you can save more than 50 tracks in trigger period, increase value. And insert `id` for playlist with ban tracks.

```js
function logFavoriteSongs() {
  const FILENAME = 'ArchiveFavoriteSongs';
  const COUNT_TRACKS = 50;

  let banTracks = Source.getPlaylistTracks('', 'id');
  let archiveTracks = Cache.read(FILENAME);
  let currentTracks = Source.getSavedTracks(archiveTracks.length > 0 ? COUNT_TRACKS : undefined);

  Combiner.push(archiveTracks, currentTracks);
  Filter.dedupTracks(archiveTracks);
  Filter.removeTracks(archiveTracks, banTracks);
  Cache.write(FILENAME, archiveTracks);

  Library.checkFavoriteTracks(archiveTracks);
  let obj = archiveTracks.reduce((obj, track) => {
    track.isFavorite ? obj.saved.push(track) : obj.unsaved.push(track);
    return obj;
  }, { saved: [], unsaved: [] })

  Order.sort(archiveTracks, 'meta.added_at', 'asc');
  Order.sort(obj.unsaved, 'meta.added_at', 'asc');
  Order.sort(obj.saved, 'meta.added_at', 'asc');

  save('All favorite', archiveTracks);
  save('Removed favorite', obj.unsaved);
  save('Present favorite', obj.saved);

  function save(name, tracks, id) {
    Playlist.saveWithUpdate({
      //id: id,
      name: name,
      tracks: tracks,
      randomCover: 'update',
      public: false,
      toEnd: true,
    });
  }
}
```

**Reply by ultrlight** [reply]
Thank you for your reply, this is working perfectly! 

I have a previous archive playlist of 4000 tracks which I combined with the 'All favourite' playlist. I have also sorted the tracks and created another playlist called 'DeleteFromArchives'. This playlist is for tracks that I accidentally like, so I do not want to write them to the file. I'm not very experienced with code, but this is what I have written :

```javascript
function logFavoriteSongs(){
  const FILENAME = 'ArchiveFavoriteSongs';
  const COUNT_TRACKS = 50;
  let archiveTracks = Cache.read(FILENAME);
  let currentTracks = Source.getSavedTracks(archiveTracks.length > 0 ? COUNT_TRACKS : undefined);
  Combiner.push(archiveTracks, currentTracks);
  Filter.dedupTracks(archiveTracks);
  Cache.write(FILENAME, archiveTracks);
  
  Library.checkFavoriteTracks(archiveTracks);
  let obj = archiveTracks.reduce((obj, track) =>{
    track.isFavorite ? obj.saved.push(track) : obj.unsaved.push(track);
    return obj;
  }, { saved: [], unsaved: [] })

// Combines my old archive playlist with the 'All favourite' playlist
  let oldArchiveTracks = Source.getPlaylistTracks('archive','abcdef')
  Combiner.push(archiveTracks,oldArchiveTracks)
  Filter.dedupTracks(archiveTracks)
  Cache.write(FILENAME, archiveTracks);

// Removes tracks which I accidentally like from the file
  let banTracks = Source.getTracks([
  {name: 'DeleteFromArchives', id: 'abcdef'},
  ]);
  Cache.remove(FILENAME, banTracks);

// Sorts tracks into order added
  Order.sort(obj.unsaved, 'meta.added_at', 'asc');
  Order.sort(obj.saved, 'meta.added_at', 'asc');
  Order.sort(archiveTracks, 'meta.added_at', 'asc');

  Playlist.saveWithUpdate({
    // id: '',
    name: 'All favorite',
    tracks: archiveTracks,
    randomCover: 'update',
    public: false,
    toEnd: true,
  });
  Playlist.saveWithUpdate({
    // id: '',
    name: 'Removed favorite',
    tracks: obj.unsaved,
    randomCover: 'update',
    public: false,
    toEnd: true,
  });
  Playlist.saveWithUpdate({
    //id: '',
    name: 'Present favorite',
    tracks: obj.saved,
    randomCover: 'update',
    public: false,
    toEnd: true,
  });
}
```

When I run this, only some of the banTracks are removed from the playlists. Also, is there a way to optimise it so it makes less requests?

I appreciate your help!

**Reply by Chimildic** [reply]
Do you add new tracks to `oldArchiveTracks`? Do you use `saveWithUpdate` in order to save added date?
`Cache.remove` removing file, not tracks.

**Reply by ultrlight** [reply]
My mistake, thank you for noticing this! So, if ```Cache.remove``` is incorrect, how do I fix the code to remove ```banTracks``` from all of the playlists? 

To answer your questions, ```oldArchiveTracks``` is a static playlist that I am not using anymore. I would just like to move the tracks from ```oldArchiveTracks``` to the 'All favourite' playlist. I would also like to use ```saveWithUpdate``` to save added date in order.

**Reply by Chimildic** [reply]
`oldArchiveTracks` need to playlist `removed favorite`?

**Reply by ultrlight** [reply]
Yes please, I would like to add the favourite songs which I removed from ```oldArchiveTracks``` to the ```Removed favourites``` playlist.

**Reply by Chimildic** [reply]
Run the next function once. The main function was edited on post.
```
function runOnce(){
  let currentTracks = Source.getSavedTracks();
  let oldTracks = Source.getPlaylistTracks('', 'id');
  Combiner.push(currentTracks, oldTracks);
  Filter.dedupTracks(currentTracks);
  Order.sort(currentTracks, 'meta.added_at', 'asc');
  Cache.write('ArchiveFavoriteSongs', currentTracks);
}
```

**Reply by ultrlight** [reply]
I've tested it and it's working perfectly, thank you so much for your help! :)
